File: leetcode/matrix/54. Spiral Matrix/cpp/main.cpp
// Time O(N^2)
// Space O(N^2)
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if (matrix.size() == 0) return {};
        
        int n = matrix.size();
        int m = matrix[0].size();
        vector<int> res;
        int startRow = 0;
        int endRow = n - 1;
        int startCol = 0;
        int endCol = m - 1;

        while (startRow <= endRow && startCol <= endCol) {
            // right
            for(int i = startCol; i <= endCol; i++) {
                res.push_back(matrix[startRow][i]);
            }
            startRow++;
            
            // down
            for(int i = startRow; i <= endRow; i++) {
                res.push_back(matrix[i][endCol]);
            }
            endCol--;
            
            // left
            if (endRow >= startRow) {
                for(int i = endCol; i >= startCol; i--) {
                    res.push_back(matrix[endRow][i]);
                }
            }
            endRow--;
            
            // up
            if (endCol >= startCol) {
                for(int i = endRow; i >= startRow; i--) {
                    res.push_back(matrix[i][startCol]);
                }
            }
            startCol++;
        
        }
        
        return res;
    }
};

File: leetcode/matrix/54. Spiral Matrix/js/main.js
// Time O(N^2)
// Space O(N^2)
const spiralOrder = matrix => {
  if (matrix.length == 0) return [];
  let n = matrix.length;
  let m = matrix[0].length;

  let startRow = 0;
  let endRow = n - 1;

  let startCol = 0;
  let endCol = m - 1;

  let res = [];

  while (startRow <= endRow && startCol <= endCol) {
    // right
    for (let i = startCol; i <= endCol; i++) {
      res.push(matrix[startRow][i]);
    }
    startRow++;

    // down
    for (let i = startRow; i <= endRow; i++) {
      res.push(matrix[i][endCol]);
    }
    endCol--;

    // left
    if (endRow >= startRow) {
      for (let i = endCol; i >= startCol; i--) {
        res.push(matrix[endRow][i]);
      }
    }
    endRow--;

    // up
    if (endCol >= startCol) {
      for (let i = endRow; i >= startRow; i--) {
        res.push(matrix[i][startCol]);
      }
    }
    startCol++;
  }

  return res;
};
